File: swift-bridge/Sources/EventKitBridge/Participant.swift
import EventKit
import Foundation

enum EKRParticipantStatus: String, Codable {
    case unknown
    case pending
    case accepted
    case declined
    case tentative
    case delegated
    case completed
    case inProcess
}

enum EKRParticipantRole: String, Codable {
    case unknown
    case required
    case optional
    case chair
    case nonParticipant
}

enum EKRParticipantScheduleStatus: String, Codable {
    case none
    case pending
    case sent
    case delivered
    case recipientNotRecognized
    case noPrivileges
    case deliveryFailed
    case cannotDeliver
    case recipientNotAllowed
}

enum EKRParticipantType: String, Codable {
    case unknown
    case person
    case room
    case resource
    case group
}

struct EKRParticipantPayload: Codable {
    var url: String?
    var name: String?
    var participantStatus: EKRParticipantStatus
    var participantRole: EKRParticipantRole
    var participantType: EKRParticipantType
    var isCurrentUser: Bool
    var contactPredicate: String?
}

func ekrParticipantStatusPayload(from status: EKParticipantStatus) -> EKRParticipantStatus {
    switch status {
    case .unknown:
        return .unknown
    case .pending:
        return .pending
    case .accepted:
        return .accepted
    case .declined:
        return .declined
    case .tentative:
        return .tentative
    case .delegated:
        return .delegated
    case .completed:
        return .completed
    case .inProcess:
        return .inProcess
    @unknown default:
        return .unknown
    }
}

func ekrParticipantRolePayload(from role: EKParticipantRole) -> EKRParticipantRole {
    switch role {
    case .unknown:
        return .unknown
    case .required:
        return .required
    case .optional:
        return .optional
    case .chair:
        return .chair
    case .nonParticipant:
        return .nonParticipant
    @unknown default:
        return .unknown
    }
}

func ekrParticipantTypePayload(from type: EKParticipantType) -> EKRParticipantType {
    switch type {
    case .unknown:
        return .unknown
    case .person:
        return .person
    case .room:
        return .room
    case .resource:
        return .resource
    case .group:
        return .group
    @unknown default:
        return .unknown
    }
}

func ekrEncodeParticipant(_ participant: EKParticipant) -> EKRParticipantPayload {
    EKRParticipantPayload(
        url: participant.url.absoluteString,
        name: participant.name,
        participantStatus: ekrParticipantStatusPayload(from: participant.participantStatus),
        participantRole: ekrParticipantRolePayload(from: participant.participantRole),
        participantType: ekrParticipantTypePayload(from: participant.participantType),
        isCurrentUser: participant.isCurrentUser,
        contactPredicate: participant.contactPredicate.predicateFormat
    )
}

File: src/participant.rs
//! EventKit participant snapshots.

use serde::{Deserialize, Serialize};

use crate::private::redacted;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
/// Represents the EventKit participant status.
pub enum EKParticipantStatus {
    #[default]
    /// Matches the EventKit `unknown` case.
    Unknown,
    /// Matches the EventKit `pending` case.
    Pending,
    /// Matches the EventKit `accepted` case.
    Accepted,
    /// Matches the EventKit `declined` case.
    Declined,
    /// Matches the EventKit `tentative` case.
    Tentative,
    /// Matches the EventKit `delegated` case.
    Delegated,
    /// Matches the EventKit `completed` case.
    Completed,
    /// Matches the EventKit `inProcess` case.
    InProcess,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
/// Represents the EventKit participant role.
pub enum EKParticipantRole {
    #[default]
    /// Matches the EventKit `unknown` case.
    Unknown,
    /// Matches the EventKit `required` case.
    Required,
    /// Matches the EventKit `optional` case.
    Optional,
    /// Matches the EventKit `chair` case.
    Chair,
    /// Matches the EventKit `nonParticipant` case.
    NonParticipant,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
/// Represents the EventKit participant schedule-delivery status.
pub enum EKParticipantScheduleStatus {
    #[default]
    /// Matches the EventKit `none` case.
    None,
    /// Matches the EventKit `pending` case.
    Pending,
    /// Matches the EventKit `sent` case.
    Sent,
    /// Matches the EventKit `delivered` case.
    Delivered,
    /// Matches the EventKit `recipientNotRecognized` case.
    RecipientNotRecognized,
    /// Matches the EventKit `noPrivileges` case.
    NoPrivileges,
    /// Matches the EventKit `deliveryFailed` case.
    DeliveryFailed,
    /// Matches the EventKit `cannotDeliver` case.
    CannotDeliver,
    /// Matches the EventKit `recipientNotAllowed` case.
    RecipientNotAllowed,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
/// Represents the EventKit participant type.
pub enum EKParticipantType {
    #[default]
    /// Matches the EventKit `unknown` case.
    Unknown,
    /// Matches the EventKit `person` case.
    Person,
    /// Matches the EventKit `room` case.
    Room,
    /// Matches the EventKit `resource` case.
    Resource,
    /// Matches the EventKit `group` case.
    Group,
}

#[derive(Clone, PartialEq, Eq, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
/// Represents EventKit participant data.
pub struct EKParticipant {
    /// Mirrors the EventKit `url` property.
    pub url: Option<String>,
    /// Mirrors the EventKit `name` property.
    pub name: Option<String>,
    /// Mirrors the EventKit `participantStatus` property.
    pub participant_status: EKParticipantStatus,
    /// Mirrors the EventKit `participantRole` property.
    pub participant_role: EKParticipantRole,
    /// Mirrors the EventKit `participantType` property.
    pub participant_type: EKParticipantType,
    /// Mirrors the EventKit `isCurrentUser` property.
    pub is_current_user: bool,
    /// Mirrors the EventKit `contactPredicate` property.
    pub contact_predicate: Option<String>,
}

impl core::fmt::Debug for EKParticipant {
    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
        f.debug_struct("EKParticipant")
            .field("url", &redacted(self.url.as_ref()))
            .field("name", &redacted(self.name.as_ref()))
            .field("participant_status", &self.participant_status)
            .field("participant_role", &self.participant_role)
            .field("participant_type", &self.participant_type)
            .field("is_current_user", &self.is_current_user)
            .field(
                "contact_predicate",
                &redacted(self.contact_predicate.as_ref()),
            )
            .finish()
    }
}
